src/Tools/VMC/RecordList.h

```diff
@@ -0,0 +1,100 @@
+////////////////////////////////////////////////////////////////////////
+// \author Christopher Backhouse -- c.backhouse@ucl.ac.uk
+////////////////////////////////////////////////////////////////////////
+
+#ifndef _RECORDLIST_H
+#define _RECORDLIST_H
+
+#include <vector>
+#include <string>
+
+class TFile;
+class TTree;
+
+namespace genie{
+namespace vmc{
+  //---------------------------------------------------------------------------
+  struct Particle
+  {
+    int pdg;
+    float E, px, py, pz;
+  };
+
+  //---------------------------------------------------------------------------
+  struct Record
+  {
+    Record();
+    Record(float _E, /*float _w,*/ int _prod_id, const std::vector<Particle>& _ps);
+
+    /// Order by energy
+    bool operator<(const Record& rhs) const;
+
+    float E;
+    //    float weight;
+    int prod_id;
+    std::vector<Particle> parts;
+  };
+
+  //---------------------------------------------------------------------------
+  class IRecordList
+  {
+  public:
+    virtual ~IRecordList(){}
+    virtual const Record* GetRecord(float E) const = 0;
+  };
+
+  //---------------------------------------------------------------------------
+  /// Helper for \ref SimpleRecordList and \ref OnDemandRecordList
+  class RecordLoader
+  {
+  public:
+    RecordLoader(TTree* tree);//const std::string& fname, const std::string& trName);
+    ~RecordLoader();
+
+    long NRecords() const;
+    Record GetRecord(int i) const;
+  protected:
+    TTree* fTree;
+
+    float Enu;//, weight;
+    int prod_id;
+    int nparts;
+    int pdgs[1024];
```

Comment:
fixed memory? Why cannot we use vector?

Reply:
This is the target of a SetBranchAddress(). There's no way to know the number of particles up-front. A vector that's fixed to 1024 elements isn't really better. std::array is marginally better, but probably not available in old C++. I think it's fine to just set some extremely high upper limit and trust the event libraries to keep within it.

Reply:
Added a comment.